File: ios/AnkiAnywhere/AnkiAnywhere/Services/APIService.swift
import Foundation

class APIService {
    static let shared = APIService()
    
    private let baseURL: String
    private var authToken: String?
    
    private init() {
        self.baseURL = UserDefaults.standard.string(forKey: "apiBaseURL") ?? "http://localhost:3000/api"
        self.authToken = UserDefaults.standard.string(forKey: "authToken")
    }
    
    func setAuthToken(_ token: String) {
        self.authToken = token
        UserDefaults.standard.set(token, forKey: "authToken")
    }
    
    func setBaseURL(_ url: String) {
        UserDefaults.standard.set(url, forKey: "apiBaseURL")
    }
    
    private func request<T: Decodable>(
        endpoint: String,
        method: String = "GET",
        body: Data? = nil,
        completion: @escaping (Result<T, Error>) -> Void
    ) {
        guard let url = URL(string: "\(baseURL)\(endpoint)") else {
            completion(.failure(NSError(domain: "Invalid URL", code: -1, userInfo: nil)))
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        if let body = body {
            request.httpBody = body
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data else {
                completion(.failure(NSError(domain: "No data", code: -1, userInfo: nil)))
                return
            }
            
            do {
                let decoder = JSONDecoder()
                decoder.dateDecodingStrategy = .iso8601
                let result = try decoder.decode(T.self, from: data)
                completion(.success(result))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }
    
    // Deck APIs
    func getDecks(completion: @escaping (Result<[Deck], Error>) -> Void) {
        request(endpoint: "/decks", completion: completion)
    }
    
    func createDeck(name: String, description: String?, completion: @escaping (Result<Deck, Error>) -> Void) {
        let body = ["name": name, "description": description ?? ""]
        guard let jsonData = try? JSONSerialization.data(withJSONObject: body) else {
            completion(.failure(NSError(domain: "JSON Error", code: -1, userInfo: nil)))
            return
        }
        request(endpoint: "/decks", method: "POST", body: jsonData, completion: completion)
    }
    
    // Card APIs
    func getCards(deckId: Int, completion: @escaping (Result<[Card], Error>) -> Void) {
        request(endpoint: "/cards/deck/\(deckId)", completion: completion)
    }
    
    func getRandomCards(deckId: Int, limit: Int = 10, completion: @escaping (Result<[Card], Error>) -> Void) {
        request(endpoint: "/cards/deck/\(deckId)/random?limit=\(limit)", completion: completion)
    }
    
    // Review APIs
    func reviewCard(cardId: Int, quality: Int, completion: @escaping (Result<CardReview, Error>) -> Void) {
        let body = ["cardId": cardId, "quality": quality] as [String : Any]
        guard let jsonData = try? JSONSerialization.data(withJSONObject: body) else {
            completion(.failure(NSError(domain: "JSON Error", code: -1, userInfo: nil)))
            return
        }
        request(endpoint: "/reviews/review", method: "POST", body: jsonData, completion: completion)
    }
    
    func getDueCards(deckId: Int? = nil, completion: @escaping (Result<[Card], Error>) -> Void) {
        let endpoint = deckId != nil ? "/reviews/due?deckId=\(deckId!)" : "/reviews/due"
        request(endpoint: endpoint, completion: completion)
    }
}

File: android/app/src/main/java/com/ankianywhere/services/APIService.kt
package com.ankianywhere.services

import android.content.Context
import android.content.SharedPreferences
import com.ankianywhere.models.*
import com.google.gson.GsonBuilder
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.util.Date

class APIService(context: Context) {
    private val prefs: SharedPreferences = context.getSharedPreferences("anki_anywhere", Context.MODE_PRIVATE)
    private val client = OkHttpClient()
    private val gson = GsonBuilder()
        .setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
        .create()
    
    private var baseURL: String
        get() = prefs.getString("apiBaseURL", "http://localhost:3000/api") ?: "http://localhost:3000/api"
        set(value) = prefs.edit().putString("apiBaseURL", value).apply()
    
    private var authToken: String?
        get() = prefs.getString("authToken", null)
        set(value) = prefs.edit().putString("authToken", value).apply()
    
    fun setAuthToken(token: String) {
        authToken = token
    }
    
    fun setBaseURL(url: String) {
        baseURL = url
    }
    
    private fun buildRequest(endpoint: String, method: String = "GET", body: RequestBody? = null): Request {
        val url = "$baseURL$endpoint"
        val builder = Request.Builder().url(url)
        
        authToken?.let {
            builder.addHeader("Authorization", "Bearer $it")
        }
        
        builder.addHeader("Content-Type", "application/json")
        
        when (method) {
            "POST" -> builder.post(body ?: "".toRequestBody())
            "PUT" -> builder.put(body ?: "".toRequestBody())
            "DELETE" -> builder.delete()
            else -> builder.get()
        }
        
        return builder.build()
    }
    
    // Deck APIs
    fun getDecks(callback: (Result<List<Deck>>) -> Unit) {
        val request = buildRequest("/decks")
        
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback(Result.failure(e))
            }
            
            override fun onResponse(call: Call, response: Response) {
                response.body?.string()?.let { responseBody ->
                    try {
                        val decks = gson.fromJson(responseBody, Array<Deck>::class.java).toList()
                        callback(Result.success(decks))
                    } catch (e: Exception) {
                        callback(Result.failure(e))
                    }
                } ?: callback(Result.failure(Exception("Empty response")))
            }
        })
    }
    
    // Card APIs
    fun getCards(deckId: Int, callback: (Result<List<Card>>) -> Unit) {
        val request = buildRequest("/cards/deck/$deckId")
        
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback(Result.failure(e))
            }
            
            override fun onResponse(call: Call, response: Response) {
                response.body?.string()?.let { responseBody ->
                    try {
                        val cards = gson.fromJson(responseBody, Array<Card>::class.java).toList()
                        callback(Result.success(cards))
                    } catch (e: Exception) {
                        callback(Result.failure(e))
                    }
                } ?: callback(Result.failure(Exception("Empty response")))
            }
        })
    }
    
    fun getRandomCards(deckId: Int, limit: Int = 10, callback: (Result<List<Card>>) -> Unit) {
        val request = buildRequest("/cards/deck/$deckId/random?limit=$limit")
        
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback(Result.failure(e))
            }
            
            override fun onResponse(call: Call, response: Response) {
                response.body?.string()?.let { responseBody ->
                    try {
                        val cards = gson.fromJson(responseBody, Array<Card>::class.java).toList()
                        callback(Result.success(cards))
                    } catch (e: Exception) {
                        callback(Result.failure(e))
                    }
                } ?: callback(Result.failure(Exception("Empty response")))
            }
        })
    }
    
    // Review APIs
    fun reviewCard(cardId: Int, quality: Int, callback: (Result<CardReview>) -> Unit) {
        val json = JSONObject()
        json.put("cardId", cardId)
        json.put("quality", quality)
        
        val body = json.toString().toRequestBody("application/json".toMediaType())
        val request = buildRequest("/reviews/review", "POST", body)
        
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback(Result.failure(e))
            }
            
            override fun onResponse(call: Call, response: Response) {
                response.body?.string()?.let { responseBody ->
                    try {
                        val review = gson.fromJson(responseBody, CardReview::class.java)
                        callback(Result.success(review))
                    } catch (e: Exception) {
                        callback(Result.failure(e))
                    }
                } ?: callback(Result.failure(Exception("Empty response")))
            }
        })
    }
    
    fun getDueCards(deckId: Int? = null, callback: (Result<List<Card>>) -> Unit) {
        val endpoint = if (deckId != null) "/reviews/due?deckId=$deckId" else "/reviews/due"
        val request = buildRequest(endpoint)
        
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback(Result.failure(e))
            }
            
            override fun onResponse(call: Call, response: Response) {
                response.body?.string()?.let { responseBody ->
                    try {
                        val cards = gson.fromJson(responseBody, Array<Card>::class.java).toList()
                        callback(Result.success(cards))
                    } catch (e: Exception) {
                        callback(Result.failure(e))
                    }
                } ?: callback(Result.failure(Exception("Empty response")))
            }
        })
    }
}
